text/0000-glimmer-component-signature.md

```diff
@@ -0,0 +1,361 @@
+
+This RFC proposes a change to that type parameter to become `Signature`, capturing more complete information about how components can be used in a template, including their expected **arguments**, the **block parameters** they provide, and what type of **element(s)** they apply any received attributes and modifiers to.
+
+This RFC is based in large part on prior work by [@gossi] and on learnings from [Glint].
+
+[@gossi]: https://github.com/gossi
+[Glint]: https://github.com/typed-ember/glint
+
+## Motivation
+
+When a developer goes to invoke a component in a template, there are a variety of questions about its interface they need to be able to answer in order to determine how to use it correctly.
+
+1. What arguments does this component accept?
+   - Which arguments are required and which are optional?
+   - What does each argument do?
+   - What types of values am I expected to pass for a given argument?
+2. How can I nest content within this component?
+   - Does it accept a default block? Named blocks?
+   - In what context will my block content appear in the component's output?
+   - For any blocks I write, what parameters does the component expose to them?
+3. In what ways can I treat this component like a regular HTML/SVG element?
+   - Can I pass attributes to it?
+   - If I apply a modifier, what kind(s) of DOM `Element` object will the modifier see?
+
+In the Glimmer `Component` base class today, a TypeScript component author can answer the first bucket of questions above by defining an `Args` type for the component class. Using this structured documentation, developer tooling can surface information about a component's arguments to consumers without requiring them to go read either ad-hoc prose or the implementation itself to determine the component's intended usage.
+
+For the second and third sets of questions, however, there is no such affordance for providing structured answers.
+
+The goal of this change is to enable tooling to provide a more complete picture for developers of the ways in which Glimmer components can be used. The word "tooling" here is used in broad terms, including:
+ - API/design system documentation generators, like [Storybook] or [Hokulea]
+ - Language servers that provide hover information, go-to-definition, etc. like [TypeScript] or [uELS]
+ - Template-aware static analysis and typechecking systems, like [Glint] or [`els-addon-typed-templates`]
+
+[Storybook]: https://storybook.js.org/
+[Hokulea]: https://github.com/hokulea/hokulea
+[TypeScript]: https://www.typescriptlang.org/
+[uELS]: https://github.com/lifeart/ember-language-server
+[Glint]: https://github.com/typed-ember/glint
+[`els-addon-typed-templates`]: https://github.com/lifeart/els-addon-typed-templates
+
+## Detailed design
+
+This RFC proposes that a component written like this today:
```

Comment:
Actually, it says "Detailed Design" and the first thing I see is the _old/existing_ design. I'd rather want to read what this new _Signature_ is about. Maybe like this:

```ts
import Component from '@glimmer/component';

export interface GreetingSignature {
  Args: {...};
  BlockParams: {...};
  Element: {...};
}

export default class Greeting extends Component<GreetingArgs> {
  // ...
}
```

and then say, `Args`, `BlockParams` and `Element` are then explained in detail below. Move the `Args` example to `Args`?

Reply:
I think it's important that this RFC isn't proposing a new green field addition, but rather a breaking change to an existing API. The proposal here is the _change_, and why we think that change is worthwhile, which is why I think it's key to present both the _from_ and the _to_.

That said, I'll see if I can condense both the before and the after here to better highlight the actual change.

Reply:
The intent was totally readable... yet as a reader I was more interessted in the new design first and then how to the migration is happening ;)